### PR TITLE
Aarnq/thanos swift support

### LINKED ISSUE #1200: [3] thanos - add swift support
**Is your feature request related to a problem? Please describe.**
We would like to be able to use s3 swift with thanos, where is supported.

**Definition of done:**
- is using swift with thanos feasible?
- make the necessary changes in the repo to support that
- update the templates is necessary

### PR DESCRIPTION
**Pull request by aarnq**
**What this PR does / why we need it**:

Adds Swift support for Thanos, and I updated the way we set it for Harbor to be more up to date and understandable.

**Which issue this PR fixes**:
Fixes #1200

**Special notes for reviewer**:
~~I still have some network policy issues to deal with.~~
And I created a PR upstream to enable auth with application credentials.

**Checklist:**

- [x] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [x] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [x] is completely transparent, will not impact the workload in any way.
    - [x] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [ ] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.
- Chart checklist (pick exactly one):
    - [x] I upgraded no Chart.
    - [ ] I upgraded a Chart and determined that no migration steps are needed.
    - [ ] I upgraded a Chart and added [migration steps](https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration).

**Pipeline config** *(if applicable)*
If you change some config options (e.g. add/rename variable or change the default value) you may need to update the config used by the pipeline in `pipeline/config`.

### CODE REVIEW

File: bin/init.bash [hunk 1]
```diff
@@ -279,15 +279,15 @@ set_harbor_config() {
...
-            yq4 --inplace '.harbor.persistence.swift.identityApiVersion = 3' "${file}"
-            yq4 --inplace '.harbor.persistence.swift.authURL = "set-me"' "${file}"
-            yq4 --inplace '.harbor.persistence.swift.regionName = "set-me"' "${file}"
-            yq4 --inplace '.harbor.persistence.swift.projectDomainName = "set-me"' "${file}"
-            yq4 --inplace '.harbor.persistence.swift.userDomainName = "set-me"' "${file}"
-            yq4 --inplace '.harbor.persistence.swift.projectName = "set-me"' "${file}"
-            yq4 --inplace '.harbor.persistence.swift.projectID = "set-me"' "${file}"
-            yq4 --inplace '.harbor.persistence.swift.tenantName = "set-me"' "${file}"
-            yq4 --inplace '.harbor.persistence.swift.authVersion = 3' "${file}"
+            yq4 --inplace '.objectStorage.swift.authVersion = 0' "${file}"
```
Comment (Xartos) [inline bin/init.bash:282]: Why do we want to use 0?

File: bin/update-ips.bash [hunk 2]
```diff
@@ -206,6 +242,26 @@ if [[ "${CHECK_CLUSTER}" =~ ^(wc|both)$ ]]; then
...
+if [[ "${CHECK_CLUSTER}" =~ ^(sc|both)$ ]]; then
+    CHECK="$(yq_dig 'sc' '.harbor.persistence.type == "swift" or .thanos.objectStorage.type == "swift"' 'false')"
+    if [ "$CHECK" == "true" ]; then
+        SWIFT_ENDPOINT="$(yq_dig 'sc' '.objectStorage.swift.authUrl' '""' | sed 's/https\?:\/\///' | sed 's/[:\/].*//')"
+        if [ -z "$SWIFT_ENDPOINT" ]; then
+            log_error "No Swift endpoint found, check your sc-config.yaml"
+            exit 1
+        fi
+        SWIFT_PORT="$(yq_dig 'sc' '.objectStorage.swift.authUrl' '""' | sed 's/https\?:\/\///' | sed 's/[A-Za-z.:]*//' | sed 's/\/.*//')"
+        if [ -z "$SWIFT_ENDPOINT" ]; then
```
Comment (Xartos) [inline bin/update-ips.bash:255]:
```suggestion
        if [ -z "$SWIFT_PORT" ]; then
```

File: helmfile/values/thanos/objectstorage-secret.yaml.gotmpl [hunk 3]
```diff
@@ -1,20 +1,49 @@
...
     secret_key: {{ .Values.objectStorage.s3.secretKey }}
     region: {{ .Values.objectStorage.s3.region }}
+
+  {{- else if eq $type "swift" }}
+  {{- $swift := .Values.objectStorage | getOrNil "swift" | required "Swift enabled for Thanos but not configured!" }}
+  type: swift
+  config:
+    auth_version: {{ $swift.authVersion }}
+    auth_url: {{ $swift.authUrl }}
+    {{- if hasKey $swift "region" }}
```
Comment (Xartos) [inline helmfile/values/thanos/objectstorage-secret.yaml.gotmpl:29]: Do we want to add these extra checks that we have here to the harbor config as well?

File: migration/v0.26.x-v0.27.x/migrate-swift.sh [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+config_common="$CK8S_CONFIG_PATH/common-config.yaml"
+config_sc="$CK8S_CONFIG_PATH/sc-config.yaml"
+secrets="$CK8S_CONFIG_PATH/secrets.yaml"
+
+configs=( "$default_common" "$default_sc" "$config_common" "$config_sc" )
+
+if [ ! -f "$secrets" ]; then
+    echo "err: $secrets not found" >&2
+    exit 1
+  fi
```
Comment (Xartos) [inline migration/v0.26.x-v0.27.x/migrate-swift.sh:18]:
```suggestion
  echo "err: $secrets not found" >&2
  exit 1
fi
```

File: migration/v0.26.x-v0.27.x/migrate-swift.sh [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+for config in "${configs[@]}"; do
+  if [ ! -f "$config" ]; then
+    echo "err: $config not found" >&2
+    exit 1
+  fi
+done
+
+merge="$(yq4 ea "explode(.) as \$config ireduce ({}; . * \$config)" "${configs[@]}")"
```
Comment (Xartos) [inline migration/v0.26.x-v0.27.x/migrate-swift.sh:27]: Wow, I assume that this basically does the override config thing? That's pretty compact

File: migration/v0.26.x-v0.27.x/migrate-swift.sh [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+if [ "$(echo "$merge" | yq4 '.harbor.persistence.type')" != "swift" ]; then
+  echo "skip: swift not enabled for harbor"
```
Comment (Xartos) [inline migration/v0.26.x-v0.27.x/migrate-swift.sh:30]: Should you stop here if this is the case?
```suggestion
  echo "skip: swift not enabled for harbor"
  exit 0
```

File: bin/init.bash (hunk 1 above)
Comment (aarnq) [reply bin/init.bash:282]: Auto detect in ncw/swift.

File: helmfile/values/thanos/objectstorage-secret.yaml.gotmpl (hunk 3 above)
Comment (aarnq) [reply helmfile/values/thanos/objectstorage-secret.yaml.gotmpl:29]: Some are a requirement for harbor but not for thanos/objstore.

Comment (Xartos) [reply helmfile/values/thanos/objectstorage-secret.yaml.gotmpl:29]: Right :+1:

**Review by Xartos — APPROVED** [review]

**Review by crssnd — APPROVED** [review]
didn't test it but look good

File: migration/v0.26.x-v0.27.x/migrate-swift.sh (hunk 5 above)
Comment (aarnq) [reply migration/v0.26.x-v0.27.x/migrate-swift.sh:27]: Yeah, I think it should be possible to grab a specific key this way as well by piping the explode and grabbing the key you want before the reduce.

**Pull request merged**